### PR TITLE
Bug 1761689 - Allow searching for XPIDL constants in C++ code

### PR DESCRIPTION
**Pull request by bill-mccloskey**
The Searchfox XPIDL analyzer doesn't match up IDL enums with their C++ equivalents. At the time I wrote the IDL analysis, the C++ indexer didn't support enums. This patch fixes the problem.

The C++ analysis uses the mangled name "E_<...>_ENUM_NAME" for an enum constant ENUM_NAME. The ... part is a hash that's supposed to be unique to the enum type. The IDL analysis needs to figure out this mangled name for each IDL constant.

To do this, it looks at the JSON data generated by the C++ analysis for the IDL file's C++ header. It looks for any definition of something of the form "E_..._CONSTANT" for CONSTANT. It assumes this is a match. Technically, if you have two classes defined in the same IDL file, and they use constants with the same name, it will get confused. Oh well.

### CODE REVIEW

**Comment by amccreight** [comment]
Sorry for the delay. I lost track of this. Above you said "The ... part is a hash that's supposed to be unique to the enum type." In my local run, ... has a value like "<T_mozilla::dom::IDBRequestReadyState>", which isn't really what I think of as a "hash function", but I guess it works. Is that expected?

**Comment by amccreight** [comment]
I'm not sure if my review comments necessarily actually need changes or not.

File: idl-analyze.py [hunk 1]
```diff
@@ -56,75 +56,93 @@ def read_cpp_analysis(fname):
     try:
         lines = open(p).readlines()
     except IOError, e:
-        return None
-    decls = {}
+        sys.exit(0)
```
[the remark this answers is not in the document]
Comment (bill-mccloskey) [reply idl-analyze.py]: Not all IDL files are actually built by the build system, so sometimes there isn't a corresponding analysis file. I'll add a comment.

File: idl-analyze.py [hunk 2]
```diff
@@ -56,75 +56,93 @@ def read_cpp_analysis(fname):
...
     for line in lines:
         j = json.loads(line.strip())
-        if 'target' in j and j['kind'] == 'decl' and j['sym'].startswith('_Z'):
-            idents = parse_mangled(j['sym'])
-            if idents and len(idents) == 2:
-                decls.setdefault(idents[0], {})[idents[1]] = j['sym']
-    return decls
-
-def method_name(m):
+        if 'target' in j and j['kind'] in ('decl', 'def'):
```
[the remark this answers is not in the document]
Comment (bill-mccloskey) [reply idl-analyze.py]: Methods are only declared in an IDL .h file, but IDL constants are defined.

File: idl-analyze.py [hunk 3]
```diff
@@ -135,109 +153,51 @@ def handle_interface(analysis, iface):
...
-                'sym': mangled_getter,
-            }
-            print json.dumps(j)
+                target(lineno, colno, 'method', m.name, '#' + m.name)
 
-            if not m.readonly:
-                mangled_setter = analysis[setter_name(m)]
+        elif isinstance(m, xpidl.Attribute):
+            mangled_getter = methods[cxx_getter_name(m)]
+            mangled_setter = methods.get(cxx_setter_name(m))
```
[the remark this answers is not in the document]
Comment (bill-mccloskey) [reply idl-analyze.py]: Because the setter won't be defined if it's a readonly attribute. It was just easier to factor the code this way.

**Comment by asutherland** [comment]
I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1761689 for de-bitrotting this functionality and adding test coverage; more comments on the bug.

**Comment by asutherland** [comment]
I landed the core of the logic from this PR in https://github.com/mozsearch/mozsearch/pull/603 so I'm closing this.

**Pull request closed without merging**